File: 117. Populating Next Right Pointers in Each Node II.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None
        root.next = None
        previous_stack = [root]
        while previous_stack:
            current_stack = []
            for _ in range(len(previous_stack)):
                current = previous_stack.pop(0)
                if current.left:
                    if current_stack:
                        current_stack[-1].next = current.left
                    current_stack.append(current.left)
                if current.right:
                    if current_stack:
                        current_stack[-1].next = current.right
                    current_stack.append(current.right)
            previous_stack = current_stack[:]
        return root            
                
            
                
        return root
                

File: 117. Populating Next Right Pointers in Each Node II.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null){
            return null;
        }
        List<Node> previous = new ArrayList<>();
        previous.add(root);
        while(!previous.isEmpty()){
            List<Node> current_row = new ArrayList<>();
            for(int index = 0; index < previous.size(); index++){
                Node current = previous.get(index);
                if(current.left != null){
                    current_row.add(current.left);
                }
                if(current.right != null){
                    current_row.add(current.right);
                }
            }
            for(int index = 0; index < current_row.size() - 1; index++){
                current_row.get(index).next = current_row.get(index+1);
            }
            previous = current_row;
        }
        return root;
    }
}